src/Cli/dotnet/commands/dotnet-new/PostActions/DotnetAddJsonPropertyPostActionProcessor.cs

```diff
@@ -0,0 +1,235 @@
+
+            if (jsonContent == null)
+            {
+                return null;
+            }
+
+            JsonNode? parentProperty = FindJsonNode(jsonContent, propertyPath, propertyPathSeparator);
+
+            if (parentProperty == null)
+            {
+                Reporter.Error.WriteLine(LocalizableStrings.PostAction_ModifyJson_Error_ParentPropertyPathInvalid);
+                return null;
+            }
+
+            try
+            {
+                parentProperty[newJsonPropertyName] = JsonNode.Parse(newJsonPropertyValue);
+            }
+            catch (JsonException)
+            {
+                parentProperty[newJsonPropertyName] = newJsonPropertyValue;
+            }
+
+            return jsonContent;
+        }
+
+        private static JsonNode? FindJsonNode(JsonNode content, string? nodePath, string pathSeparator)
+        {
+            if (nodePath == null)
+            {
+                return content;
+            }
+
+            string[] properties = nodePath.Split(pathSeparator);
+
+            JsonNode? node = content;
+
+            foreach (string property in properties)
+            {
+                if (node == null)
+                {
+                    return null;
+                }
+
+                node = node[property];
+            }
+
+            return node;
+        }
+
+        private static IReadOnlyList<string> FindFilesAtOrAbovePath(
+            IPhysicalFileSystem fileSystem,
+            string startPath,
+            string matchPattern,
+            Func<string, bool>? secondaryFilter = null,
+            int maxAllowedAboveDirectories = 250)
+        {
+            string? directory = fileSystem.DirectoryExists(startPath) ? startPath : Path.GetDirectoryName(startPath);
+
+            if (directory == null)
+            {
+                throw new InvalidOperationException();
+            }
+
+            int numberOfUpLevels = 0;
+
+            do
+            {
+                List<string> filesInDir = fileSystem.EnumerateFileSystemEntries(directory, matchPattern, SearchOption.AllDirectories).ToList();
+                List<string> matches = new();
+
+                matches = secondaryFilter == null ? filesInDir : filesInDir.Where(x => secondaryFilter(x)).ToList();
+
+                if (matches.Count > 0)
+                {
+                    return matches;
+                }
+
+                directory = Path.GetPathRoot(directory) != directory ? Directory.GetParent(directory)?.FullName : null;
+                numberOfUpLevels++;
+            }
+            while (directory != null && numberOfUpLevels <= maxAllowedAboveDirectories);
+
+            return new List<string>();
+        }
+
+        private static (string sourceName, string replaceValue)? DetermineSourceNameReplaceValue(ICreationEffects creationEffects)
```

Comment:
`sourceName` is not limited to project file only.
I would recommend to consider all file renames when replacing args.

Reply:
Is this related to your other [comment](https://github.com/dotnet/sdk/pull/29880/files/227889fb859648b37f026cfd31c83e5eb8bcacaa#r1072102296) ?  So what you're actually saying is that I should look at every file to determine this source name & replacement value ?

Reply:
Yes, it is related. 

There is no proven way to determine what is `sourceName` having the information available for the post action. 
Relying that project file is changed by `sourceName` is not fully correct, though it might be so for many cases. But since this is a generic action that modifies arbitrary json file, this assumption cannot be taken as granted.

Therefore I would recommend to consider either replacing all file renames for this properties, or better create an additional configuration which file renames to consider. This configuration should contain the list of source filenames, check the final names for those and apply renames to `newJsonPropertyName` and `newJsonPropertyValue`. 

We can improve API, so source name is known for post action context, but that a separate problem to solve.

Reply:
> We can improve API, so source name is known for post action context, but that a separate problem to solve.

I think this would be the best and most clear solution where everyone can benefit from.  However, I do realise this might not be something that can be fixed on short term.

Reply:
> Therefore I would recommend to consider either replacing all file renames for this properties, or better create an additional configuration which file renames to consider. This configuration should contain the list of source filenames, check the final names for those and apply renames to `newJsonPropertyName` and `newJsonPropertyValue`.
> 

So, If I understand correctly, you would go over all the FileChanges that are listed in the `CreationEffects` and take all those changes into account for applying renames to `newJsonProperty` and `newJsonPropertyValue`?
I would assume though that most files will not be renamed.  Only files that contain the `sourceName` 'magic value' will be renamed.
Or you mean that I should not just assume that the project file would contain the `sourceName` value ?  It looks to me that this file has the most change of being renamed, but indeed, I can take into consideration all files.

Reply:
> So, If I understand correctly, you would go over all the FileChanges that are listed in the CreationEffects and take all those changes into account for applying renames to newJsonProperty and newJsonPropertyValue?
I would assume though that most files will not be renamed. Only files that contain the sourceName 'magic value' will be renamed.

The files can be quite frequently renamed:
- `sourceName` as one of the case, likely most frequent
- `fileRename` configuration
- `source` - `rename` configuration

Assuming that project name is `sourceName` is not correct: every occurrence of `sourceName` is replaced, and project name may be different from it.

> But additionally, what is a file has the name sourceNameSomethingElse. This will be replaced to MyReplaceValueSomethingElse. We also need to be able to cope with that.

That's a valid concern, but mostly it's up to template author to ensure that's not happening. The replacement value for `sourceName` might be something very unique: as `MyValueForSourceNameThatCannotOccurSomewhereElse.1`, then accidental replacements are pretty safe.

Reply:
I've been thinkering about this for a while, and I'm actually convinced that the only proper way to correctly handle this, is to make sure that the `sourceName` and the value that should eventually be used to replace  `sourceName` should actually be available in the `PostActionProcessor`  instance.

I think this information can be part of the `ICreationEffects2` interface.

Reply:
Hey @fgheysels, just checking if you are continue working on this issue! We are very interested in this solution and are willing to help once we have more time. Let us know the situation so we can proceed in the best way :)

Reply:
Hi @maridematte, thanks for reaching out.

I'm actually struggling a bit with how we could make sure that we take into account the `sourceName` replacement in the JSON property that we could add.

Let's say for instance that we want to add a JSON property to an existing JSON file, but the name of that property or the value of that property should be 'dynamic'. 
The templates support the `sourceName` property and when instantiating the template, the template-files that contain the value of `sourceName` are being replaced by the string that is specified by the `--name`  argument.
I want to have this behavior when specifying the new JSON property or JSON value as well.  Suppose:

```json
"sourceName":  "TheMagicWord"

"postActions": [
    {
        "args": {
               "newJsonPropertyName":  "TheMagicWord"
              ...
        }
    }
]
```
When instantiating the template and specifying `--name MyProjectName`, I would like to have that the new property that must be added to the json file is not `TheMagicWord` , but is `MyProjectName`.

For now, I've implemented this by checking the `creationEffects`  on the project-file as this is afaik the only way to get to know the original `sourceName` value and the value of the `--name` argument (see [here](https://github.com/dotnet/sdk/blob/497187db31346639dfc1e390256e64dbd4b3a7e7/src/Cli/Microsoft.TemplateEngine.Cli/PostActionProcessors/AddJsonPropertyPostActionProcessor.cs#L204).   However, @vlada-shubina  has a relevant remark on this [here](https://github.com/dotnet/sdk/pull/29880/#discussion_r1072100716), as there is indeed an assumption there that the name of the project file in the template is actually the same as the `sourceName` value.  It is true that we should not assume that, and that I need to look at other files as well.
Next to that, I also assume that the project file will not have a suffix or prefix to the `sourceName`.  (It can actually be `TheMagicWordSomeProject.csproj` for instance.
So actually, it would be better if I could somehow calculate the difference between the `SourceRelativePath` and `TargetRelativePath` of the creation-effects, but I don't see how this can be done.

As said in my comment just above yours, I think the best solution would be to extend the `ICreationEffects2` interface with the value of `sourceName` and the value of the `--name` argument that might have been given.

Reply:
It would be good to think about scenarios other then `sourceName` here, as this is not only source of replacement.
I think https://github.com/dotnet/templating/issues/1600 would be beneficial in general if extended to all post action arguments, however unfortunately it's not on the radar yet. 

If this is implemented the example below:
```
  "sourceName": "MyProject",
  "postActions": [
    {
      "description": "Modify JSON file",
      "manualInstructions": [ { "text": "Modify the JSON file manually." } ],
      "args": {
        "jsonFileName": "deployment.template.json",
        "parentPropertyPath": "modulesContent:$edgeAgent:properties.desired:modules",
        "newJsonPropertyName": "MyProject",
        "newJsonPropertyValue": "{\"version\":\"1.0.0\", \"type\":\"docker\", \"settings\":{\"image\":\"${MODULEDIR<../MyProject>}\"}}"
      },
      "actionId": "695A3659-EB40-4FF5-A6A6-C9C4E629FCB0",
      "continueOnError": true
    }

```
will result in having `MyProject` replaced with whatever `sourceName` is in args collection and post action won't need to care about this.

Another suggestion might be to extend interface of post actions with actual replacements or operations doing replacements but it might be too heavy.

Having just `sourceName` in interface or replacing just `sourceName` might be an intermediate solution but we need to aim for full support further on.

Reply:
I'm just wondering now how to proceed with this.
It seems like we're stuck with this for now, as we're waiting on some other functionality as mentioned by @vlada-shubina to be able to fully complete the implementation of this.

Reply:
We are currently discussing how to proceed with it. 
One option on the table is to implement https://github.com/dotnet/templating/issues/1600 in near future (1-2 months). 
This PR may not wait for it, it will start to work for replacements once https://github.com/dotnet/templating/issues/1600 is merged or changed if needed then.

We are definitely interested in merging this implementation even without possibility to replace source name or other replacements. I know unfortunately it doesn't work for your case now, but we see value even without replacement for other uses.

Please allow us more time to discuss it.

Reply:
Hi @fgheysels,

We have prioritized the work that has to be done in order to unblock you.
It is expected to implement within the next 1-2 months. You can track the progress in this ticket https://github.com/dotnet/templating/issues/1600